Currency providers and OrderPart currency

MatteoPiovanelli-Laser opened the pull request:
I added the interface for currency selection discussed in #99 as well as the currency in the OrderPart as per #102 .

The default `ISelectedCurrencyProvider` in the Nwazet.Commerce feature replicates the original behaviour of using the Site's culture for currency.
Anothe `ISelectedCurrencyProvider` is in its own feature and uses a Site setting to select the currency to be used across the whole store.

There were changes in several controllers to inject the active ISelectedCurrencyProvider, as well as changes in views to use it in formatting prices. I also added the `Currency` service to give some static formatting methods.

I added a "Settings" submenu under "Commerce" for SIteSettings strictly related to the store. The one setting like that I implemented is for the currency provider that uses a Site Setting to set the currency used (it is in its own feature).

Adding the CurrencyCode to the OrderPart caused some changes that are not retrocompatible, to populate that value when creating the order.
For new orders, the currency code will be populated (in the changes I made, when paying with stripe I use the Stripe currency).
For pre-existing orders, that property is not populated: on displaying/editing an order, I show the prices using the current currency (as per the active provider) but I don't change the information in the order, since there is no guarantee that the current currency is the correct one.

As things are now, it is still possible to set a currency for payments (Stripe here) that is different than the one used to display prices, and there is nothing special happening when the currency is changed: the absolute value of prices and such does not change.

MatteoPiovanelli-Laser commented:
@bleroy @petedavis 
How does this look?

bleroy commented on Services/ISelectedCurrencyProvider.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <summary>
+        /// The name of the provider is used to identify uniquely the provider.
+        /// </summary>
+        string Name { get; }
```
Once we have a currency class, we can return an instance of that instead of having separate methods and properties to get name, symbol, etc.

MatteoPiovanelli-Laser replied:
Yes on the currency class, but this `Name` here would be the name of the provider, not of the currency

bleroy replied:
Right, never mind for name then, but my point remains for the others.

bleroy commented on Services/Currency.cs:
```diff
@@ -0,0 +1,256 @@
+        /// <example>Given "EUR", this will return "€".</example>
+        public static string GetCurrencyFormat(string code) {
+            //http://www.thefinancials.com/?SubSectionID=curformat has a table including formats
+            //TODO: implement symbols from that table
```
That should also be on the currency class.

bleroy replied:
Also, I'd rather not have TODOs before I merge the PR.

MatteoPiovanelli-Laser replied:
I will take care of this.

bleroy commented on Services/Currency.cs:
```diff
@@ -0,0 +1,256 @@
+        }
+
+        public static string GetPriceAsString(double price, string code) {
+            return price.ToString(GetCurrencyFormat(code)) + " " + GetCurrencySymbol(code);
```
This is wrong: the currency symbol is not always after the amount. For instance, $400.

bleroy commented on Services/UseCurrencyFromSiteSettingsProvider.cs:
```diff
@@ -0,0 +1,63 @@
+        public string GetPriceString(decimal price) {
+            return price.ToString(GetCurrencyFormat()) + " " + GetCurrencySymbol();
+        }
+        public string GetPriceString(double? price) {
```
See above.

bleroy commented:
I also think we should have tests associated with those changes.

MatteoPiovanelli-Laser commented on Menus/ECommerceSettingsAdminMenu.cs:
```diff
@@ -0,0 +1,41 @@
+                    .Position("2")
+                    .LinkToFirstChild(false)
+
+                    .Add(subItem => subItem
```
Done

MatteoPiovanelli-Laser commented on Models/ECommerceCurrencySiteSettingsPart.cs:
```diff
@@ -0,0 +1,18 @@
+        {
+            get {
+                string cc = this.Retrieve(p => p.CurrencyCode);
+                return string.IsNullOrWhiteSpace(cc) ? "USD" : cc; //default is USD
```
I feel that using the site's culture is a rather clunky way of initializing this site's setting with a default value. Also a `System.Globalization.NumberFormatInfo` object doe not seem to contain a currency's 3-letters code, but only its symbol.

bleroy replied:
Arguably, defaulting on USD is worse...

bleroy replied:
Yes, I see what you're saying. It's not like defaulting a culture. OK, then maybe this should throw.

bleroy replied:
Yes, this should definitely throw.

MatteoPiovanelli-Laser commented on Services/ISelectedCurrencyProvider.cs:
```diff
@@ -0,0 +1,73 @@
+using Orchard;
+
+namespace Nwazet.Commerce.Services {
+    public interface ISelectedCurrencyProvider : IDependency {
```
ok

MatteoPiovanelli-Laser commented on Services/Currency.cs:
```diff
@@ -0,0 +1,256 @@
+            "ZWD Zimbabwe Dollar"
+        };
+        #endregion
+        private static Lazy<Dictionary<string, string>> _currencyCodes =
```
OK

MatteoPiovanelli-Laser commented:
Implementing the currency class and the tests (insert animated work in progress icon here)

MatteoPiovanelli-Laser commented:
I think the current PR answers the changes above, except:
What tests did you have in mind? I didn't implement any yet.

MatteoPiovanelli-Laser commented:
Also,
A separate project for the tests? Or should I place them where the code is? Maybe in a block like 
```
#IF TESTS
...
#ENDIF
```

MatteoPiovanelli-Laser commented:
In case of a separate test project, I don't know how to include that in the PR/the repo

MatteoPiovanelli-Laser commented:
A further note, on the code:
In `CurrencyProviderBase.cs` you'll see I am always using the `CurrentCulture`. The idea there is that if I set my UI to Italian, I still want to see "1.234,00 $", rather than "$1,234.00" as would be the "representation" in the en-US culture. However, in the methods of the `Currency` class I am changing the currency symbol and decimal digits to match the currency being used.

MatteoPiovanelli-Laser commented:
Something annoying I noticed is that many currencies use the "$" symbol. Thta may lead to confusion unless noted.
e.g. The Argentine Peso uses the "$" sign.

bleroy commented:
Tests should go in the test project: https://github.com/bleroy/Nwazet.Commerce/tree/master/Nwazet.Commerce.Tests

bleroy commented:
> if I set my UI to Italian, I still want to see "1.234,00 $", rather than "$1,234.00" as would be the "representation" in the en-US culture.

I don't understand what you're saying here. The dollar sign should always be in front for a sum in dollars. Decimal points and commas are culture-based, yes.

bleroy commented:
>Something annoying I noticed is that many currencies use the "$" symbol. Thta may lead to confusion unless noted.
> e.g. The Argentine Peso uses the "$" sign.

That is the problem of the site administrator. We should just implement cultures and currencies as they are standardized and used.

MatteoPiovanelli-Laser commented:
> I don't understand what you're saying here. The dollar sign should always be in front for a sum in dollars. Decimal points and commas are culture-based, yes.

Actually, the position of the currency symbol is culture based as well.

MatteoPiovanelli-Laser commented on Models/OrderPart.cs:
```diff
@@ -405,7 +405,7 @@ public string CurrencyCode
                 Status + " - " +
                 BillingAddress.Honorific + " " + BillingAddress.FirstName + " " + BillingAddress.LastName + " - " +
-                Currency.GetPriceAsString(Total, CurrencyCode) +
+                CurrencyCode + " " + Total.ToString() +
```
I am trying to find a way to revert to using a `GetPriceAsString`, however proper formatting of the currency requires specifying the culture. Here, and in general in static methods like `GetPriceAsString used to be, I don't know the site's culture. I could use the InvariantCulture as default, but that may look odd, as we would end up with two potentially different formatting standards around the store.

MatteoPiovanelli-Laser replied:
off the top of my head, the best I came up with is `Currency.Currencies[CurrencyCode].PriceAsString(Total, CultureInfo.CurrentCulture)`

MatteoPiovanelli-Laser commented on Drivers/ECommerceCurrencySiteSettingsPartDriver.cs:
```diff
@@ -23,7 +26,7 @@ protected override string Prefix
         
         protected override DriverResult Editor(ECommerceCurrencySiteSettingsPart part, dynamic shapeHelper) {
-            if (_selectedCurrencyProvider is UseCurrencyFromSiteSettingsProvider) {
+            if (_currencyProvider is UseCurrencyFromSiteSettingsProvider) {
```
I see your point. You are correct. My concern here was the case when someone would be enabling a different feature giving a different provider, But even if they did, that provider would not (should not) be using the setting here.

MatteoPiovanelli-Laser commented:
About the tests:

What did you have in mind?
I was thinking about verifying that things are formatted as expected, but that feels more like I'd be testing that `ToString()` works as expected, rather than anything I have implemented here.

MatteoPiovanelli-Laser commented:
Note a thing that is left to the site's administrator to handle:

Supposing Stripe is being used for payments, the currency used there should be set to be the same as the store's, otherwise it's probably a crime of some kind.

bleroy commented on Migrations/OrderMigrations.cs:
```diff
@@ -43,5 +59,19 @@ public class OrderMigrations : DataMigrationImpl {
                 .AddColumn<int>("UserId", column => column.WithDefault(-1)));
             return 3;
```
Shouldn't that return 4 now?

MatteoPiovanelli-Laser replied:
I added the `UpdateFrom3()` method, that returns 4

MatteoPiovanelli-Laser replied:
I think I see what you mean. 
You are saying that the `UpdateFrom3()` that I added needs not run on new tenants, so this `UpdateFrom2()` might as well return 4 already. Right?

bleroy commented:
Looks good except for one small thing in the migration. Thanks!

bleroy commented:
On tests: I'd like to see verification that all non-trivial behavior is as expected, for example currency formatting, yes.

MatteoPiovanelli-Laser commented:
In my latest commits:

- I changed what gets injected into the currency providers: `IWorkContextAccessor` rather than `IOrchardServices`.
- I fixed the build-errors in the test project, related to changes in the constructors of some driver.
- I fixed a bug in the `ValidateAttribute` method of `ProductAttributesPartDriver` (see #109)
- I updated the test project so it work with O1.10: upgraded the used framework and changed the nunit references to NuGet.
- I added a few formatting tests related to the Currency class.

bleroy commented:
Looks good, thanks. Can you check the migration issue?

bleroy commented:
Thanks!
